From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcrt: Implement _dosmaperr, the FLS/TLS helpers and other CRT internals.

CRT internals that were spec stubs: _dosmaperr (errno and _doserrno from
a system error), __get_app_type (msvcrt, msvcr80/90) and ucrtbase's
_query_app_type, __p__osplatform and _get_winver, __get_tlsindex and
__get_flsindex (the CRT's thread data index), __fls_getvalue/__fls_setvalue
with the i386 ___fls_getvalue@4/___fls_setvalue@8 and __set_flsgetvalue,
_getdcwd_nolock/_wgetdcwd_nolock, _ValidateRead/_ValidateWrite/
_ValidateExecute, type_info's static _Name_base(_internal) and
_Type_info_dtor(_internal) and _type_info_dtor_internal_method,
_inconsistency (terminate), the int overloads of _set_new_handler and
_set_se_translator (reset), msvcr110/120's debug placement operator
delete/delete[] and __ExceptionPtrSwap. msvcr120_app forwards.
---
 dlls/msvcr100/msvcr100.spec         | 60 +++++++++++++--------------
 dlls/msvcr110/msvcr110.spec         | 64 ++++++++++++++---------------
 dlls/msvcr120/msvcr120.spec         | 64 ++++++++++++++---------------
 dlls/msvcr120_app/msvcr120_app.spec | 52 +++++++++++------------
 dlls/msvcr80/msvcr80.spec           | 70 +++++++++++++++----------------
 dlls/msvcr90/msvcr90.spec           | 62 ++++++++++++++--------------
 dlls/msvcrt/cpp.c                   | 26 ++++++++++++
 dlls/msvcrt/data.c                  | 38 +++++++++++++++++
 dlls/msvcrt/dir.c                   | 18 ++++++++
 dlls/msvcrt/exception_ptr.c         | 18 ++++++++
 dlls/msvcrt/heap.c                  | 24 +++++++++++
 dlls/msvcrt/misc.c                  | 82 +++++++++++++++++++++++++++++++++++++
 dlls/msvcrt/msvcrt.spec             |  2 +-
 dlls/ucrtbase/ucrtbase.spec         |  2 +-
 14 files changed, 394 insertions(+), 188 deletions(-)

diff --git a/dlls/msvcr100/msvcr100.spec b/dlls/msvcr100/msvcr100.spec
index 006f570..4c8cf15 100644
--- a/dlls/msvcr100/msvcr100.spec
+++ b/dlls/msvcr100/msvcr100.spec
@@ -459,10 +459,10 @@
 @ stub -arch=arm ?_IsCanceling@_TaskCollection@details@Concurrency@@QAA_NXZ
 @ stub -arch=i386 ?_IsCanceling@_TaskCollection@details@Concurrency@@QAE_NXZ
 @ stub -arch=win64 ?_IsCanceling@_TaskCollection@details@Concurrency@@QEAA_NXZ
-@ stub -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z
-@ stub -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z
-@ stub -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z
-@ stub -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z
+@ cdecl -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
 @ cdecl -arch=arm ?_NumberOfSpins@?$_SpinWait@$00@details@Concurrency@@IAAKXZ(ptr) SpinWait__NumberOfSpins
 @ thiscall -arch=i386 ?_NumberOfSpins@?$_SpinWait@$00@details@Concurrency@@IAEKXZ(ptr) SpinWait__NumberOfSpins
 @ cdecl -arch=win64 ?_NumberOfSpins@?$_SpinWait@$00@details@Concurrency@@IEAAKXZ(ptr) SpinWait__NumberOfSpins
@@ -550,17 +550,17 @@
 @ stub -arch=arm ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QAA_NXZ
 @ stub -arch=i386 ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QAE_NXZ
 @ stub -arch=win64 ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QEAA_NXZ
-@ stub -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z
-@ stub -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z
-@ stub -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z
-@ stub -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z
+@ cdecl -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z(ptr) type_info__Type_info_dtor
 @ cdecl -arch=win32 ?_UnderlyingYield@details@Concurrency@@YAXXZ() _UnderlyingYield
-@ stub -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z
-@ stub ?_ValidateExecute@@YAHP6GHXZ@Z
-@ stub -arch=win32 ?_ValidateRead@@YAHPBXI@Z
-@ stub -arch=win64 ?_ValidateRead@@YAHPEBXI@Z
-@ stub -arch=win32 ?_ValidateWrite@@YAHPAXI@Z
-@ stub -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z
+@ cdecl -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z(ptr) _ValidateExecute
+@ cdecl ?_ValidateExecute@@YAHP6GHXZ@Z(ptr) _ValidateExecute
+@ cdecl -arch=win32 ?_ValidateRead@@YAHPBXI@Z(ptr long) _ValidateRead
+@ cdecl -arch=win64 ?_ValidateRead@@YAHPEBXI@Z(ptr long) _ValidateRead
+@ cdecl -arch=win32 ?_ValidateWrite@@YAHPAXI@Z(ptr long) _ValidateWrite
+@ cdecl -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z(ptr long) _ValidateWrite
 @ cdecl ?_Value@_SpinCount@details@Concurrency@@SAIXZ() SpinCount__Value
 @ cdecl -arch=win32 ?__ExceptionPtrAssign@@YAXPAXPBX@Z(ptr ptr) __ExceptionPtrAssign
 @ cdecl -arch=win64 ?__ExceptionPtrAssign@@YAXPEAXPEBX@Z(ptr ptr) __ExceptionPtrAssign
@@ -579,7 +579,7 @@
 @ cdecl -arch=win32 ?__ExceptionPtrRethrow@@YAXPBX@Z(ptr) __ExceptionPtrRethrow
 @ cdecl -arch=win64 ?__ExceptionPtrRethrow@@YAXPEBX@Z(ptr) __ExceptionPtrRethrow
 @ cdecl __uncaught_exception()
-@ stub ?_inconsistency@@YAXXZ
+@ cdecl ?_inconsistency@@YAXXZ() _inconsistency
 @ cdecl -arch=win32 ?_invalid_parameter@@YAXPBG00II@Z(wstr wstr wstr long long) _invalid_parameter
 @ cdecl -arch=win64 ?_invalid_parameter@@YAXPEBG00I_K@Z(wstr wstr wstr long long) _invalid_parameter
 @ cdecl -arch=win32 ?_is_exception_typeof@@YAHABVtype_info@@PAU_EXCEPTION_POINTERS@@@Z(ptr ptr) _is_exception_typeof
@@ -592,20 +592,20 @@
 @ cdecl -arch=win32 ?_query_new_handler@@YAP6AHI@ZXZ() _query_new_handler
 @ cdecl -arch=win64 ?_query_new_handler@@YAP6AH_K@ZXZ() _query_new_handler
 @ cdecl ?_query_new_mode@@YAHXZ() _query_new_mode
-@ stub -arch=win32 ?_set_new_handler@@YAP6AHI@ZH@Z  # int (__cdecl*__cdecl _set_new_handler(int))(unsigned int)
-@ stub -arch=win64 ?_set_new_handler@@YAP6AH_K@ZH@Z  # int (__cdecl*__cdecl _set_new_handler(int))(unsigned __int64)
+@ cdecl -arch=win32 ?_set_new_handler@@YAP6AHI@ZH@Z(long) _set_new_handler_zero
+@ cdecl -arch=win64 ?_set_new_handler@@YAP6AH_K@ZH@Z(long) _set_new_handler_zero
 @ cdecl -arch=win32 ?_set_new_handler@@YAP6AHI@ZP6AHI@Z@Z(ptr) _set_new_handler
 @ cdecl -arch=win64 ?_set_new_handler@@YAP6AH_K@ZP6AH0@Z@Z(ptr) _set_new_handler
 @ cdecl ?_set_new_mode@@YAHH@Z(long) _set_new_mode
-@ stub -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZH@Z  # void (__cdecl*__cdecl _set_se_translator(int))(unsigned int,struct _EXCEPTION_POINTERS *)
-@ stub -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZH@Z  # void (__cdecl*__cdecl _set_se_translator(int))(unsigned int,struct _EXCEPTION_POINTERS * __ptr64)
+@ cdecl -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZH@Z(long) _set_se_translator_zero
+@ cdecl -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZH@Z(long) _set_se_translator_zero
 @ cdecl -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z(ptr) _set_se_translator
 @ cdecl -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z(ptr) _set_se_translator
 @ cdecl -arch=win32 ?_sopen@@YAHPBDHHH@Z(str long long long) _sopen
 @ cdecl -arch=win64 ?_sopen@@YAHPEBDHHH@Z(str long long long) _sopen
-@ stub -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ
-@ stub -arch=i386 ?_type_info_dtor_internal_method@type_info@@QAEXXZ  # public: void __thiscall type_info::_type_info_dtor_internal_method(void)
-@ stub -arch=win64 ?_type_info_dtor_internal_method@type_info@@QEAAXXZ  # public: void __cdecl type_info::_type_info_dtor_internal_method(void) __ptr64
+@ cdecl -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ(ptr) type_info_dtor
+@ thiscall -arch=i386 ?_type_info_dtor_internal_method@type_info@@QAEXXZ(ptr) type_info_dtor
+@ cdecl -arch=win64 ?_type_info_dtor_internal_method@type_info@@QEAAXXZ(ptr) type_info_dtor
 @ cdecl -arch=win32 ?_wopen@@YAHPB_WHH@Z(wstr long long) _wopen
 @ cdecl -arch=win64 ?_wopen@@YAHPEB_WHH@Z(wstr long long) _wopen
 @ cdecl -arch=win32 ?_wsopen@@YAHPB_WHHH@Z(wstr long long long) _wsopen
@@ -756,13 +756,13 @@
 @ cdecl __dllonexit(ptr ptr ptr)
 @ cdecl __doserrno()
 @ cdecl __dstbias() __p__dstbias
-@ stub ___fls_getvalue@4
-@ stub ___fls_setvalue@8
+@ stdcall ___fls_getvalue@4(long) fls_getvalue_stdcall
+@ stdcall ___fls_setvalue@8(long ptr) fls_setvalue_stdcall
 @ cdecl __fpecode()
 @ cdecl __free_locale(ptr) _free_locale
 @ cdecl __get_current_locale() _get_current_locale
-@ stub __get_flsindex
-@ stub __get_tlsindex
+@ cdecl __get_flsindex() __get_tlsindex
+@ cdecl __get_tlsindex()
 @ cdecl __getmainargs(ptr ptr ptr long ptr)
 @ extern __initenv MSVCRT___initenv
 @ cdecl __iob_func()
@@ -824,7 +824,7 @@
 @ cdecl __pxcptinfoptrs()
 @ cdecl __report_gsfailure(ptr)
 @ cdecl __set_app_type(long)
-@ stub __set_flsgetvalue
+@ cdecl __set_flsgetvalue()
 @ extern __setlc_active MSVCRT___setlc_active
 @ cdecl __setusermatherr(ptr)
 @ cdecl __strncnt(str long)
@@ -940,7 +940,7 @@
 @ extern _daylight MSVCRT___daylight
 @ cdecl _difftime32(long long)
 @ cdecl _difftime64(int64 int64)
-@ stub _dosmaperr
+@ cdecl _dosmaperr(long)
 @ extern _dstbias MSVCRT__dstbias
 @ cdecl _dup(long)
 @ cdecl _dup2(long long)
@@ -1067,7 +1067,7 @@
 @ cdecl _getche_nolock()
 @ cdecl _getcwd(str long)
 @ cdecl _getdcwd(long str long)
-@ stub _getdcwd_nolock
+@ cdecl _getdcwd_nolock(long ptr long)
 @ cdecl _getdiskfree(long ptr)
 @ cdecl _getdllprocaddr(long str long)
 @ cdecl _getdrive()
@@ -1730,7 +1730,7 @@
 @ cdecl _wfullpath(ptr wstr long)
 @ cdecl _wgetcwd(wstr long)
 @ cdecl _wgetdcwd(long wstr long)
-@ stub _wgetdcwd_nolock
+@ cdecl _wgetdcwd_nolock(long ptr long)
 @ cdecl _wgetenv(wstr)
 @ cdecl _wgetenv_s(ptr ptr long wstr)
 @ cdecl _wmakepath(ptr wstr wstr wstr wstr)
diff --git a/dlls/msvcr110/msvcr110.spec b/dlls/msvcr110/msvcr110.spec
index 0714f77..d2671a1 100644
--- a/dlls/msvcr110/msvcr110.spec
+++ b/dlls/msvcr110/msvcr110.spec
@@ -341,8 +341,8 @@
 @ cdecl -arch=win64 ??2@YAPEAX_KHPEBDH@Z(long long str long) operator_new_dbg
 @ cdecl -arch=win32 ??3@YAXPAX@Z(ptr) operator_delete
 @ cdecl -arch=win64 ??3@YAXPEAX@Z(ptr) operator_delete
-@ stub -arch=win32 ??3@YAXPAXHPBDH@Z
-@ stub -arch=win64 ??3@YAXPEAXHPEBDH@Z
+@ cdecl -arch=win32 ??3@YAXPAXHPBDH@Z(ptr long str long) operator_delete_dbg
+@ cdecl -arch=win64 ??3@YAXPEAXHPEBDH@Z(ptr long str long) operator_delete_dbg
 @ cdecl -arch=arm ??4?$_SpinWait@$00@details@Concurrency@@QAAAAV012@ABV012@@Z(ptr ptr) SpinWait_assign
 @ thiscall -arch=i386 ??4?$_SpinWait@$00@details@Concurrency@@QAEAAV012@ABV012@@Z(ptr ptr) SpinWait_assign
 @ cdecl -arch=win64 ??4?$_SpinWait@$00@details@Concurrency@@QEAAAEAV012@AEBV012@@Z(ptr ptr) SpinWait_assign
@@ -398,8 +398,8 @@
 @ cdecl -arch=win64 ??_U@YAPEAX_KHPEBDH@Z(long long str long) operator_new_dbg
 @ cdecl -arch=win32 ??_V@YAXPAX@Z(ptr) operator_delete
 @ cdecl -arch=win64 ??_V@YAXPEAX@Z(ptr) operator_delete
-@ stub -arch=win32 ??_V@YAXPAXHPBDH@Z
-@ stub -arch=win64 ??_V@YAXPEAXHPEBDH@Z
+@ cdecl -arch=win32 ??_V@YAXPAXHPBDH@Z(ptr long str long) operator_delete_dbg
+@ cdecl -arch=win64 ??_V@YAXPEAXHPEBDH@Z(ptr long str long) operator_delete_dbg
 @ cdecl -arch=win32 ?Alloc@Concurrency@@YAPAXI@Z(long) Concurrency_Alloc
 @ cdecl -arch=win64 ?Alloc@Concurrency@@YAPEAX_K@Z(long) Concurrency_Alloc
 @ cdecl ?Block@Context@Concurrency@@SAXXZ() Context_Block
@@ -553,10 +553,10 @@
 @ cdecl -arch=arm ?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QBA_NXZ(ptr) _Context_IsSynchronouslyBlocked
 @ thiscall -arch=i386 ?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QBE_NXZ(ptr) _Context_IsSynchronouslyBlocked
 @ cdecl -arch=win64 ?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QEBA_NXZ(ptr) _Context_IsSynchronouslyBlocked
-@ stub -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z
-@ stub -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z
-@ stub -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z
-@ stub -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z
+@ cdecl -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
 @ stub -arch=win32 ?_NewCollection@_AsyncTaskCollection@details@Concurrency@@SAPAV123@PAV_CancellationTokenState@23@@Z
 @ stub -arch=win64 ?_NewCollection@_AsyncTaskCollection@details@Concurrency@@SAPEAV123@PEAV_CancellationTokenState@23@@Z
 @ stub -arch=win32 ?_NewTokenState@_CancellationTokenState@details@Concurrency@@SAPAV123@XZ
@@ -672,18 +672,18 @@
 @ stub -arch=arm ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QAA_NXZ
 @ stub -arch=i386 ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QAE_NXZ
 @ stub -arch=win64 ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QEAA_NXZ
-@ stub -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z
-@ stub -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z
-@ stub -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z
-@ stub -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z
+@ cdecl -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z(ptr) type_info__Type_info_dtor
 @ cdecl ?_UnderlyingYield@details@Concurrency@@YAXXZ() _UnderlyingYield
-@ stub -arch=arm ?_ValidateExecute@@YAHP6AHXZ@Z
-@ stub -arch=i386 ?_ValidateExecute@@YAHP6GHXZ@Z
-@ stub -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z
-@ stub -arch=win32 ?_ValidateRead@@YAHPBXI@Z
-@ stub -arch=win64 ?_ValidateRead@@YAHPEBXI@Z
-@ stub -arch=win32 ?_ValidateWrite@@YAHPAXI@Z
-@ stub -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z
+@ cdecl -arch=arm ?_ValidateExecute@@YAHP6AHXZ@Z(ptr) _ValidateExecute
+@ cdecl -arch=i386 ?_ValidateExecute@@YAHP6GHXZ@Z(ptr) _ValidateExecute
+@ cdecl -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z(ptr) _ValidateExecute
+@ cdecl -arch=win32 ?_ValidateRead@@YAHPBXI@Z(ptr long) _ValidateRead
+@ cdecl -arch=win64 ?_ValidateRead@@YAHPEBXI@Z(ptr long) _ValidateRead
+@ cdecl -arch=win32 ?_ValidateWrite@@YAHPAXI@Z(ptr long) _ValidateWrite
+@ cdecl -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z(ptr long) _ValidateWrite
 @ cdecl ?_Value@_SpinCount@details@Concurrency@@SAIXZ() SpinCount__Value
 @ cdecl ?_Yield@_Context@details@Concurrency@@SAXXZ() Context_Yield
 @ cdecl -arch=win32 ?__ExceptionPtrAssign@@YAXPAXPBX@Z(ptr ptr) __ExceptionPtrAssign
@@ -702,12 +702,12 @@
 @ cdecl -arch=win64 ?__ExceptionPtrDestroy@@YAXPEAX@Z(ptr) __ExceptionPtrDestroy
 @ cdecl -arch=win32 ?__ExceptionPtrRethrow@@YAXPBX@Z(ptr) __ExceptionPtrRethrow
 @ cdecl -arch=win64 ?__ExceptionPtrRethrow@@YAXPEBX@Z(ptr) __ExceptionPtrRethrow
-@ stub -arch=win32 ?__ExceptionPtrSwap@@YAXPAX0@Z
-@ stub -arch=win64 ?__ExceptionPtrSwap@@YAXPEAX0@Z
+@ cdecl -arch=win32 ?__ExceptionPtrSwap@@YAXPAX0@Z(ptr ptr) __ExceptionPtrSwap
+@ cdecl -arch=win64 ?__ExceptionPtrSwap@@YAXPEAX0@Z(ptr ptr) __ExceptionPtrSwap
 @ cdecl -arch=win32 ?__ExceptionPtrToBool@@YA_NPBX@Z(ptr) __ExceptionPtrToBool
 @ cdecl -arch=win64 ?__ExceptionPtrToBool@@YA_NPEBX@Z(ptr) __ExceptionPtrToBool
 @ cdecl __uncaught_exception()
-@ stub ?_inconsistency@@YAXXZ
+@ cdecl ?_inconsistency@@YAXXZ() _inconsistency
 @ cdecl -arch=win32 ?_invalid_parameter@@YAXPBG00II@Z(wstr wstr wstr long long) _invalid_parameter
 @ cdecl -arch=win64 ?_invalid_parameter@@YAXPEBG00I_K@Z(wstr wstr wstr long long) _invalid_parameter
 @ cdecl -arch=win32 ?_is_exception_typeof@@YAHABVtype_info@@PAU_EXCEPTION_POINTERS@@@Z(ptr ptr) _is_exception_typeof
@@ -720,20 +720,20 @@
 @ cdecl -arch=win32 ?_query_new_handler@@YAP6AHI@ZXZ() _query_new_handler
 @ cdecl -arch=win64 ?_query_new_handler@@YAP6AH_K@ZXZ() _query_new_handler
 @ cdecl ?_query_new_mode@@YAHXZ() _query_new_mode
-@ stub -arch=win32 ?_set_new_handler@@YAP6AHI@ZH@Z  # int (__cdecl*__cdecl _set_new_handler(int))(unsigned int)
-@ stub -arch=win64 ?_set_new_handler@@YAP6AH_K@ZH@Z  # int (__cdecl*__cdecl _set_new_handler(int))(unsigned __int64)
+@ cdecl -arch=win32 ?_set_new_handler@@YAP6AHI@ZH@Z(long) _set_new_handler_zero
+@ cdecl -arch=win64 ?_set_new_handler@@YAP6AH_K@ZH@Z(long) _set_new_handler_zero
 @ cdecl -arch=win32 ?_set_new_handler@@YAP6AHI@ZP6AHI@Z@Z(ptr) _set_new_handler
 @ cdecl -arch=win64 ?_set_new_handler@@YAP6AH_K@ZP6AH0@Z@Z(ptr) _set_new_handler
 @ cdecl ?_set_new_mode@@YAHH@Z(long) _set_new_mode
-@ stub -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZH@Z  # void (__cdecl*__cdecl _set_se_translator(int))(unsigned int,struct _EXCEPTION_POINTERS *)
-@ stub -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZH@Z  # void (__cdecl*__cdecl _set_se_translator(int))(unsigned int,struct _EXCEPTION_POINTERS * __ptr64)
+@ cdecl -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZH@Z(long) _set_se_translator_zero
+@ cdecl -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZH@Z(long) _set_se_translator_zero
 @ cdecl -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z(ptr) _set_se_translator
 @ cdecl -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z(ptr) _set_se_translator
 @ cdecl -arch=win32 ?_sopen@@YAHPBDHHH@Z(str long long long) _sopen
 @ cdecl -arch=win64 ?_sopen@@YAHPEBDHHH@Z(str long long long) _sopen
-@ stub -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ
-@ stub -arch=i386 ?_type_info_dtor_internal_method@type_info@@QAEXXZ  # public: void __thiscall type_info::_type_info_dtor_internal_method(void)
-@ stub -arch=win64 ?_type_info_dtor_internal_method@type_info@@QEAAXXZ  # public: void __cdecl type_info::_type_info_dtor_internal_method(void) __ptr64
+@ cdecl -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ(ptr) type_info_dtor
+@ thiscall -arch=i386 ?_type_info_dtor_internal_method@type_info@@QAEXXZ(ptr) type_info_dtor
+@ cdecl -arch=win64 ?_type_info_dtor_internal_method@type_info@@QEAAXXZ(ptr) type_info_dtor
 @ cdecl -arch=win32 ?_wopen@@YAHPB_WHH@Z(wstr long long) _wopen
 @ cdecl -arch=win64 ?_wopen@@YAHPEB_WHH@Z(wstr long long) _wopen
 @ cdecl -arch=win32 ?_wsopen@@YAHPB_WHHH@Z(wstr long long long) _wsopen
@@ -938,8 +938,8 @@
 @ cdecl __fpecode()
 @ cdecl __free_locale(ptr) _free_locale
 @ cdecl __get_current_locale() _get_current_locale
-@ stub __get_flsindex
-@ stub __get_tlsindex
+@ cdecl __get_flsindex() __get_tlsindex
+@ cdecl __get_tlsindex()
 @ cdecl __getmainargs(ptr ptr ptr long ptr)
 @ extern __initenv MSVCRT___initenv
 @ cdecl __iob_func()
@@ -1117,7 +1117,7 @@
 @ extern _daylight MSVCRT___daylight
 @ cdecl _difftime32(long long)
 @ cdecl _difftime64(int64 int64)
-@ stub _dosmaperr
+@ cdecl _dosmaperr(long)
 @ extern _dstbias MSVCRT__dstbias
 @ cdecl _dup(long)
 @ cdecl _dup2(long long)
diff --git a/dlls/msvcr120/msvcr120.spec b/dlls/msvcr120/msvcr120.spec
index f34ee87..dd1c6fd 100644
--- a/dlls/msvcr120/msvcr120.spec
+++ b/dlls/msvcr120/msvcr120.spec
@@ -338,8 +338,8 @@
 @ cdecl -arch=win64 ??2@YAPEAX_KHPEBDH@Z(long long str long) operator_new_dbg
 @ cdecl -arch=win32 ??3@YAXPAX@Z(ptr) operator_delete
 @ cdecl -arch=win64 ??3@YAXPEAX@Z(ptr) operator_delete
-@ stub -arch=win32 ??3@YAXPAXHPBDH@Z
-@ stub -arch=win64 ??3@YAXPEAXHPEBDH@Z
+@ cdecl -arch=win32 ??3@YAXPAXHPBDH@Z(ptr long str long) operator_delete_dbg
+@ cdecl -arch=win64 ??3@YAXPEAXHPEBDH@Z(ptr long str long) operator_delete_dbg
 @ cdecl -arch=arm ??4?$_SpinWait@$00@details@Concurrency@@QAAAAV012@ABV012@@Z(ptr ptr) SpinWait_assign
 @ thiscall -arch=i386 ??4?$_SpinWait@$00@details@Concurrency@@QAEAAV012@ABV012@@Z(ptr ptr) SpinWait_assign
 @ cdecl -arch=win64 ??4?$_SpinWait@$00@details@Concurrency@@QEAAAEAV012@AEBV012@@Z(ptr ptr) SpinWait_assign
@@ -395,8 +395,8 @@
 @ cdecl -arch=win64 ??_U@YAPEAX_KHPEBDH@Z(long long str long) operator_new_dbg
 @ cdecl -arch=win32 ??_V@YAXPAX@Z(ptr) operator_delete
 @ cdecl -arch=win64 ??_V@YAXPEAX@Z(ptr) operator_delete
-@ stub -arch=win32 ??_V@YAXPAXHPBDH@Z
-@ stub -arch=win64 ??_V@YAXPEAXHPEBDH@Z
+@ cdecl -arch=win32 ??_V@YAXPAXHPBDH@Z(ptr long str long) operator_delete_dbg
+@ cdecl -arch=win64 ??_V@YAXPEAXHPEBDH@Z(ptr long str long) operator_delete_dbg
 @ cdecl -arch=win32 ?Alloc@Concurrency@@YAPAXI@Z(long) Concurrency_Alloc
 @ cdecl -arch=win64 ?Alloc@Concurrency@@YAPEAX_K@Z(long) Concurrency_Alloc
 @ cdecl ?Block@Context@Concurrency@@SAXXZ() Context_Block
@@ -544,10 +544,10 @@
 @ cdecl -arch=arm ?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QBA_NXZ(ptr) _Context_IsSynchronouslyBlocked
 @ thiscall -arch=i386 ?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QBE_NXZ(ptr) _Context_IsSynchronouslyBlocked
 @ cdecl -arch=win64 ?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QEBA_NXZ(ptr) _Context_IsSynchronouslyBlocked
-@ stub -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z
-@ stub -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z
-@ stub -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z
-@ stub -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z
+@ cdecl -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
 @ stub -arch=win32 ?_NewCollection@_AsyncTaskCollection@details@Concurrency@@SAPAV123@PAV_CancellationTokenState@23@@Z
 @ stub -arch=win64 ?_NewCollection@_AsyncTaskCollection@details@Concurrency@@SAPEAV123@PEAV_CancellationTokenState@23@@Z
 @ cdecl -arch=arm ?_NumberOfSpins@?$_SpinWait@$00@details@Concurrency@@IAAKXZ(ptr) SpinWait__NumberOfSpins
@@ -655,18 +655,18 @@
 @ stub -arch=arm ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QAA_NXZ
 @ stub -arch=i386 ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QAE_NXZ
 @ stub -arch=win64 ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QEAA_NXZ
-@ stub -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z
-@ stub -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z
-@ stub -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z
-@ stub -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z
+@ cdecl -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z(ptr) type_info__Type_info_dtor
 @ cdecl ?_UnderlyingYield@details@Concurrency@@YAXXZ() _UnderlyingYield
-@ stub -arch=arm ?_ValidateExecute@@YAHP6AHXZ@Z
-@ stub -arch=i386 ?_ValidateExecute@@YAHP6GHXZ@Z
-@ stub -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z
-@ stub -arch=win32 ?_ValidateRead@@YAHPBXI@Z
-@ stub -arch=win64 ?_ValidateRead@@YAHPEBXI@Z
-@ stub -arch=win32 ?_ValidateWrite@@YAHPAXI@Z
-@ stub -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z
+@ cdecl -arch=arm ?_ValidateExecute@@YAHP6AHXZ@Z(ptr) _ValidateExecute
+@ cdecl -arch=i386 ?_ValidateExecute@@YAHP6GHXZ@Z(ptr) _ValidateExecute
+@ cdecl -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z(ptr) _ValidateExecute
+@ cdecl -arch=win32 ?_ValidateRead@@YAHPBXI@Z(ptr long) _ValidateRead
+@ cdecl -arch=win64 ?_ValidateRead@@YAHPEBXI@Z(ptr long) _ValidateRead
+@ cdecl -arch=win32 ?_ValidateWrite@@YAHPAXI@Z(ptr long) _ValidateWrite
+@ cdecl -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z(ptr long) _ValidateWrite
 @ cdecl ?_Value@_SpinCount@details@Concurrency@@SAIXZ() SpinCount__Value
 @ cdecl ?_Yield@_Context@details@Concurrency@@SAXXZ() Context_Yield
 @ cdecl -arch=win32 ?__ExceptionPtrAssign@@YAXPAXPBX@Z(ptr ptr) __ExceptionPtrAssign
@@ -685,12 +685,12 @@
 @ cdecl -arch=win64 ?__ExceptionPtrDestroy@@YAXPEAX@Z(ptr) __ExceptionPtrDestroy
 @ cdecl -arch=win32 ?__ExceptionPtrRethrow@@YAXPBX@Z(ptr) __ExceptionPtrRethrow
 @ cdecl -arch=win64 ?__ExceptionPtrRethrow@@YAXPEBX@Z(ptr) __ExceptionPtrRethrow
-@ stub -arch=win32 ?__ExceptionPtrSwap@@YAXPAX0@Z
-@ stub -arch=win64 ?__ExceptionPtrSwap@@YAXPEAX0@Z
+@ cdecl -arch=win32 ?__ExceptionPtrSwap@@YAXPAX0@Z(ptr ptr) __ExceptionPtrSwap
+@ cdecl -arch=win64 ?__ExceptionPtrSwap@@YAXPEAX0@Z(ptr ptr) __ExceptionPtrSwap
 @ cdecl -arch=win32 ?__ExceptionPtrToBool@@YA_NPBX@Z(ptr) __ExceptionPtrToBool
 @ cdecl -arch=win64 ?__ExceptionPtrToBool@@YA_NPEBX@Z(ptr) __ExceptionPtrToBool
 @ cdecl __uncaught_exception()
-@ stub ?_inconsistency@@YAXXZ
+@ cdecl ?_inconsistency@@YAXXZ() _inconsistency
 @ cdecl -arch=win32 ?_invalid_parameter@@YAXPBG00II@Z(wstr wstr wstr long long) _invalid_parameter
 @ cdecl -arch=win64 ?_invalid_parameter@@YAXPEBG00I_K@Z(wstr wstr wstr long long) _invalid_parameter
 @ cdecl -arch=win32 ?_is_exception_typeof@@YAHABVtype_info@@PAU_EXCEPTION_POINTERS@@@Z(ptr ptr) _is_exception_typeof
@@ -703,20 +703,20 @@
 @ cdecl -arch=win32 ?_query_new_handler@@YAP6AHI@ZXZ() _query_new_handler
 @ cdecl -arch=win64 ?_query_new_handler@@YAP6AH_K@ZXZ() _query_new_handler
 @ cdecl ?_query_new_mode@@YAHXZ() _query_new_mode
-@ stub -arch=win32 ?_set_new_handler@@YAP6AHI@ZH@Z  # int (__cdecl*__cdecl _set_new_handler(int))(unsigned int)
-@ stub -arch=win64 ?_set_new_handler@@YAP6AH_K@ZH@Z  # int (__cdecl*__cdecl _set_new_handler(int))(unsigned __int64)
+@ cdecl -arch=win32 ?_set_new_handler@@YAP6AHI@ZH@Z(long) _set_new_handler_zero
+@ cdecl -arch=win64 ?_set_new_handler@@YAP6AH_K@ZH@Z(long) _set_new_handler_zero
 @ cdecl -arch=win32 ?_set_new_handler@@YAP6AHI@ZP6AHI@Z@Z(ptr) _set_new_handler
 @ cdecl -arch=win64 ?_set_new_handler@@YAP6AH_K@ZP6AH0@Z@Z(ptr) _set_new_handler
 @ cdecl ?_set_new_mode@@YAHH@Z(long) _set_new_mode
-@ stub -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZH@Z  # void (__cdecl*__cdecl _set_se_translator(int))(unsigned int,struct _EXCEPTION_POINTERS *)
-@ stub -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZH@Z  # void (__cdecl*__cdecl _set_se_translator(int))(unsigned int,struct _EXCEPTION_POINTERS * __ptr64)
+@ cdecl -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZH@Z(long) _set_se_translator_zero
+@ cdecl -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZH@Z(long) _set_se_translator_zero
 @ cdecl -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z(ptr) _set_se_translator
 @ cdecl -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z(ptr) _set_se_translator
 @ cdecl -arch=win32 ?_sopen@@YAHPBDHHH@Z(str long long long) _sopen
 @ cdecl -arch=win64 ?_sopen@@YAHPEBDHHH@Z(str long long long) _sopen
-@ stub -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ
-@ stub -arch=i386 ?_type_info_dtor_internal_method@type_info@@QAEXXZ  # public: void __thiscall type_info::_type_info_dtor_internal_method(void)
-@ stub -arch=win64 ?_type_info_dtor_internal_method@type_info@@QEAAXXZ  # public: void __cdecl type_info::_type_info_dtor_internal_method(void) __ptr64
+@ cdecl -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ(ptr) type_info_dtor
+@ thiscall -arch=i386 ?_type_info_dtor_internal_method@type_info@@QAEXXZ(ptr) type_info_dtor
+@ cdecl -arch=win64 ?_type_info_dtor_internal_method@type_info@@QEAAXXZ(ptr) type_info_dtor
 @ cdecl -arch=win32 ?_wopen@@YAHPB_WHH@Z(wstr long long) _wopen
 @ cdecl -arch=win64 ?_wopen@@YAHPEB_WHH@Z(wstr long long) _wopen
 @ cdecl -arch=win32 ?_wsopen@@YAHPB_WHHH@Z(wstr long long long) _wsopen
@@ -929,8 +929,8 @@
 @ cdecl __fpecode()
 @ cdecl __free_locale(ptr) _free_locale
 @ cdecl __get_current_locale() _get_current_locale
-@ stub __get_flsindex
-@ stub __get_tlsindex
+@ cdecl __get_flsindex() __get_tlsindex
+@ cdecl __get_tlsindex()
 @ cdecl __getmainargs(ptr ptr ptr long ptr)
 @ extern __initenv MSVCRT___initenv
 @ cdecl __iob_func()
@@ -1110,7 +1110,7 @@
 @ cdecl _dclass(double)
 @ cdecl _difftime32(long long)
 @ cdecl _difftime64(int64 int64)
-@ stub _dosmaperr
+@ cdecl _dosmaperr(long)
 @ cdecl _dpcomp(double double)
 @ cdecl _dsign(double)
 @ extern _dstbias MSVCRT__dstbias
diff --git a/dlls/msvcr120_app/msvcr120_app.spec b/dlls/msvcr120_app/msvcr120_app.spec
index 246ff0a..d5ed921 100644
--- a/dlls/msvcr120_app/msvcr120_app.spec
+++ b/dlls/msvcr120_app/msvcr120_app.spec
@@ -336,8 +336,8 @@
 @ cdecl -arch=win64 ??2@YAPEAX_KHPEBDH@Z(long long str long) msvcr120.??2@YAPEAX_KHPEBDH@Z
 @ cdecl -arch=win32 ??3@YAXPAX@Z(ptr) msvcr120.??3@YAXPAX@Z
 @ cdecl -arch=win64 ??3@YAXPEAX@Z(ptr) msvcr120.??3@YAXPEAX@Z
-@ stub -arch=win32 ??3@YAXPAXHPBDH@Z
-@ stub -arch=win64 ??3@YAXPEAXHPEBDH@Z
+@ cdecl -arch=win32 ??3@YAXPAXHPBDH@Z(ptr long str long) msvcr120.??3@YAXPAXHPBDH@Z
+@ cdecl -arch=win64 ??3@YAXPEAXHPEBDH@Z(ptr long str long) msvcr120.??3@YAXPEAXHPEBDH@Z
 @ cdecl -arch=arm ??4?$_SpinWait@$00@details@Concurrency@@QAAAAV012@ABV012@@Z(ptr ptr) msvcr120.??4?$_SpinWait@$00@details@Concurrency@@QAAAAV012@ABV012@@Z
 @ thiscall -arch=i386 ??4?$_SpinWait@$00@details@Concurrency@@QAEAAV012@ABV012@@Z(ptr ptr) msvcr120.??4?$_SpinWait@$00@details@Concurrency@@QAEAAV012@ABV012@@Z
 @ cdecl -arch=win64 ??4?$_SpinWait@$00@details@Concurrency@@QEAAAEAV012@AEBV012@@Z(ptr ptr) msvcr120.??4?$_SpinWait@$00@details@Concurrency@@QEAAAEAV012@AEBV012@@Z
@@ -393,8 +393,8 @@
 @ cdecl -arch=win64 ??_U@YAPEAX_KHPEBDH@Z(long long str long) msvcr120.??_U@YAPEAX_KHPEBDH@Z
 @ cdecl -arch=win32 ??_V@YAXPAX@Z(ptr) msvcr120.??_V@YAXPAX@Z
 @ cdecl -arch=win64 ??_V@YAXPEAX@Z(ptr) msvcr120.??_V@YAXPEAX@Z
-@ stub -arch=win32 ??_V@YAXPAXHPBDH@Z
-@ stub -arch=win64 ??_V@YAXPEAXHPEBDH@Z
+@ cdecl -arch=win32 ??_V@YAXPAXHPBDH@Z(ptr long str long) msvcr120.??_V@YAXPAXHPBDH@Z
+@ cdecl -arch=win64 ??_V@YAXPEAXHPEBDH@Z(ptr long str long) msvcr120.??_V@YAXPEAXHPEBDH@Z
 @ cdecl -arch=win32 ?Alloc@Concurrency@@YAPAXI@Z(long) msvcr120.?Alloc@Concurrency@@YAPAXI@Z
 @ cdecl -arch=win64 ?Alloc@Concurrency@@YAPEAX_K@Z(long) msvcr120.?Alloc@Concurrency@@YAPEAX_K@Z
 @ cdecl ?Block@Context@Concurrency@@SAXXZ() msvcr120.?Block@Context@Concurrency@@SAXXZ
@@ -541,10 +541,10 @@
 @ cdecl -arch=arm ?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QBA_NXZ(ptr) msvcr120.?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QBA_NXZ
 @ thiscall -arch=i386 ?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QBE_NXZ(ptr) msvcr120.?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QBE_NXZ
 @ cdecl -arch=win64 ?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QEBA_NXZ(ptr) msvcr120.?_IsSynchronouslyBlocked@_Context@details@Concurrency@@QEBA_NXZ
-@ stub -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z
-@ stub -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z
-@ stub -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z
-@ stub -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z
+@ cdecl -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) msvcr120.?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z
+@ cdecl -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) msvcr120.?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z
+@ cdecl -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) msvcr120.?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z
+@ cdecl -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) msvcr120.?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z
 @ stub -arch=win32 ?_NewCollection@_AsyncTaskCollection@details@Concurrency@@SAPAV123@PAV_CancellationTokenState@23@@Z
 @ stub -arch=win64 ?_NewCollection@_AsyncTaskCollection@details@Concurrency@@SAPEAV123@PEAV_CancellationTokenState@23@@Z
 @ cdecl -arch=arm ?_NumberOfSpins@?$_SpinWait@$00@details@Concurrency@@IAAKXZ(ptr) msvcr120.?_NumberOfSpins@?$_SpinWait@$00@details@Concurrency@@IAAKXZ
@@ -652,18 +652,18 @@
 @ stub -arch=arm ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QAA_NXZ
 @ stub -arch=i386 ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QAE_NXZ
 @ stub -arch=win64 ?_TryAcquireWrite@_ReaderWriterLock@details@Concurrency@@QEAA_NXZ
-@ stub -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z
-@ stub -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z
-@ stub -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z
-@ stub -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z
+@ cdecl -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z(ptr) msvcr120.?_Type_info_dtor@type_info@@CAXPAV1@@Z
+@ cdecl -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z(ptr) msvcr120.?_Type_info_dtor@type_info@@CAXPEAV1@@Z
+@ cdecl -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z(ptr) msvcr120.?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z
+@ cdecl -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z(ptr) msvcr120.?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z
 @ cdecl ?_UnderlyingYield@details@Concurrency@@YAXXZ() msvcr120.?_UnderlyingYield@details@Concurrency@@YAXXZ
-@ stub -arch=arm ?_ValidateExecute@@YAHP6AHXZ@Z
-@ stub -arch=i386 ?_ValidateExecute@@YAHP6GHXZ@Z
-@ stub -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z
-@ stub -arch=win32 ?_ValidateRead@@YAHPBXI@Z
-@ stub -arch=win64 ?_ValidateRead@@YAHPEBXI@Z
-@ stub -arch=win32 ?_ValidateWrite@@YAHPAXI@Z
-@ stub -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z
+@ cdecl -arch=arm ?_ValidateExecute@@YAHP6AHXZ@Z(ptr) msvcr120.?_ValidateExecute@@YAHP6AHXZ@Z
+@ cdecl -arch=i386 ?_ValidateExecute@@YAHP6GHXZ@Z(ptr) msvcr120.?_ValidateExecute@@YAHP6GHXZ@Z
+@ cdecl -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z(ptr) msvcr120.?_ValidateExecute@@YAHP6A_JXZ@Z
+@ cdecl -arch=win32 ?_ValidateRead@@YAHPBXI@Z(ptr long) msvcr120.?_ValidateRead@@YAHPBXI@Z
+@ cdecl -arch=win64 ?_ValidateRead@@YAHPEBXI@Z(ptr long) msvcr120.?_ValidateRead@@YAHPEBXI@Z
+@ cdecl -arch=win32 ?_ValidateWrite@@YAHPAXI@Z(ptr long) msvcr120.?_ValidateWrite@@YAHPAXI@Z
+@ cdecl -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z(ptr long) msvcr120.?_ValidateWrite@@YAHPEAXI@Z
 @ cdecl ?_Value@_SpinCount@details@Concurrency@@SAIXZ() msvcr120.?_Value@_SpinCount@details@Concurrency@@SAIXZ
 @ cdecl ?_Yield@_Context@details@Concurrency@@SAXXZ() msvcr120.?_Yield@_Context@details@Concurrency@@SAXXZ
 @ cdecl -arch=win32 ?__ExceptionPtrAssign@@YAXPAXPBX@Z(ptr ptr) msvcr120.?__ExceptionPtrAssign@@YAXPAXPBX@Z
@@ -682,12 +682,12 @@
 @ cdecl -arch=win64 ?__ExceptionPtrDestroy@@YAXPEAX@Z(ptr) msvcr120.?__ExceptionPtrDestroy@@YAXPEAX@Z
 @ cdecl -arch=win32 ?__ExceptionPtrRethrow@@YAXPBX@Z(ptr) msvcr120.?__ExceptionPtrRethrow@@YAXPBX@Z
 @ cdecl -arch=win64 ?__ExceptionPtrRethrow@@YAXPEBX@Z(ptr) msvcr120.?__ExceptionPtrRethrow@@YAXPEBX@Z
-@ stub -arch=win32 ?__ExceptionPtrSwap@@YAXPAX0@Z
-@ stub -arch=win64 ?__ExceptionPtrSwap@@YAXPEAX0@Z
+@ cdecl -arch=win32 ?__ExceptionPtrSwap@@YAXPAX0@Z(ptr ptr) msvcr120.?__ExceptionPtrSwap@@YAXPAX0@Z
+@ cdecl -arch=win64 ?__ExceptionPtrSwap@@YAXPEAX0@Z(ptr ptr) msvcr120.?__ExceptionPtrSwap@@YAXPEAX0@Z
 @ cdecl -arch=win32 ?__ExceptionPtrToBool@@YA_NPBX@Z(ptr) msvcr120.?__ExceptionPtrToBool@@YA_NPBX@Z
 @ cdecl -arch=win64 ?__ExceptionPtrToBool@@YA_NPEBX@Z(ptr) msvcr120.?__ExceptionPtrToBool@@YA_NPEBX@Z
 @ cdecl __uncaught_exception() msvcr120.__uncaught_exception
-@ stub ?_inconsistency@@YAXXZ
+@ cdecl ?_inconsistency@@YAXXZ() msvcr120.?_inconsistency@@YAXXZ
 @ cdecl -arch=win32 ?_invalid_parameter@@YAXPBG00II@Z(wstr wstr wstr long long) msvcr120.?_invalid_parameter@@YAXPBG00II@Z
 @ cdecl -arch=win64 ?_invalid_parameter@@YAXPEBG00I_K@Z(wstr wstr wstr long long) msvcr120.?_invalid_parameter@@YAXPEBG00I_K@Z
 @ cdecl -arch=win32 ?_is_exception_typeof@@YAHABVtype_info@@PAU_EXCEPTION_POINTERS@@@Z(ptr ptr) msvcr120.?_is_exception_typeof@@YAHABVtype_info@@PAU_EXCEPTION_POINTERS@@@Z
@@ -711,7 +711,7 @@
 @ cdecl -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z(ptr) msvcr120.?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z
 @ cdecl -arch=win32 ?_sopen@@YAHPBDHHH@Z(str long long long) msvcr120.?_sopen@@YAHPBDHHH@Z
 @ cdecl -arch=win64 ?_sopen@@YAHPEBDHHH@Z(str long long long) msvcr120.?_sopen@@YAHPEBDHHH@Z
-@ stub -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ
+@ cdecl -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ(ptr) msvcr120.?_type_info_dtor_internal_method@type_info@@QAAXXZ
 @ stub -arch=i386 ?_type_info_dtor_internal_method@type_info@@QAEXXZ  # public: void __thiscall type_info::_type_info_dtor_internal_method(void)
 @ stub -arch=win64 ?_type_info_dtor_internal_method@type_info@@QEAAXXZ  # public: void __cdecl type_info::_type_info_dtor_internal_method(void) __ptr64
 @ cdecl -arch=win32 ?_wopen@@YAHPB_WHH@Z(wstr long long) msvcr120.?_wopen@@YAHPB_WHH@Z
@@ -906,8 +906,8 @@
 @ cdecl __fpecode() msvcr120.__fpecode
 @ cdecl __free_locale(ptr) msvcr120.__free_locale
 @ cdecl __get_current_locale() msvcr120.__get_current_locale
-@ stub __get_flsindex
-@ stub __get_tlsindex
+@ cdecl __get_flsindex() msvcr120.__get_flsindex
+@ cdecl __get_tlsindex() msvcr120.__get_tlsindex
 @ cdecl __getmainargs(ptr ptr ptr long ptr) msvcr120.__getmainargs
 @ extern __initenv msvcr120.__initenv
 @ cdecl __iob_func() msvcr120.__iob_func
@@ -1050,7 +1050,7 @@
 @ cdecl _dclass(double) msvcr120._dclass
 @ cdecl _difftime32(long long) msvcr120._difftime32
 @ cdecl _difftime64(int64 int64) msvcr120._difftime64
-@ stub _dosmaperr
+@ cdecl _dosmaperr(long) msvcr120._dosmaperr
 @ cdecl _dpcomp(double double) msvcr120._dpcomp
 @ cdecl _dsign(double) msvcr120._dsign
 @ extern _dstbias msvcr120._dstbias
diff --git a/dlls/msvcr80/msvcr80.spec b/dlls/msvcr80/msvcr80.spec
index 859de41..05ad394 100644
--- a/dlls/msvcr80/msvcr80.spec
+++ b/dlls/msvcr80/msvcr80.spec
@@ -85,22 +85,22 @@
 @ cdecl -arch=win64 ??_U@YAPEAX_KHPEBDH@Z(long long str long) operator_new_dbg
 @ cdecl -arch=win32 ??_V@YAXPAX@Z(ptr) operator_delete
 @ cdecl -arch=win64 ??_V@YAXPEAX@Z(ptr) operator_delete
-@ stub -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z  # private: static char const * __cdecl type_info::_Name_base(class type_info const *,struct __type_info_node *)
-@ stub -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z  # private: static char const * __ptr64 __cdecl type_info::_Name_base(class type_info const * __ptr64,struct __type_info_node * __ptr64)
-@ stub -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z  # private: static char const * __cdecl type_info::_Name_base_internal(class type_info const *,struct __type_info_node *)
-@ stub -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z  # private: static char const * __ptr64 __cdecl type_info::_Name_base_internal(class type_info const * __ptr64,struct __type_info_node * __ptr64)
-@ stub -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z  # private: static void __cdecl type_info::_Type_info_dtor(class type_info *)
-@ stub -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z  # private: static void __cdecl type_info::_Type_info_dtor(class type_info * __ptr64)
-@ stub -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z  # private: static void __cdecl type_info::_Type_info_dtor_internal(class type_info *)
-@ stub -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z  # private: static void __cdecl type_info::_Type_info_dtor_internal(class type_info * __ptr64)
-@ stub -arch=win32 ?_ValidateExecute@@YAHP6GHXZ@Z  # int __cdecl _ValidateExecute(int (__stdcall*)(void))
-@ stub -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z  # int __cdecl _ValidateExecute(__int64 (__cdecl*)(void))
-@ stub -arch=win32 ?_ValidateRead@@YAHPBXI@Z  # int __cdecl _ValidateRead(void const *,unsigned int)
-@ stub -arch=win64 ?_ValidateRead@@YAHPEBXI@Z  # int __cdecl _ValidateRead(void const * __ptr64,unsigned int)
-@ stub -arch=win32 ?_ValidateWrite@@YAHPAXI@Z  # int __cdecl _ValidateWrite(void *,unsigned int)
-@ stub -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z  # int __cdecl _ValidateWrite(void * __ptr64,unsigned int)
+@ cdecl -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win32 ?_ValidateExecute@@YAHP6GHXZ@Z(ptr) _ValidateExecute
+@ cdecl -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z(ptr) _ValidateExecute
+@ cdecl -arch=win32 ?_ValidateRead@@YAHPBXI@Z(ptr long) _ValidateRead
+@ cdecl -arch=win64 ?_ValidateRead@@YAHPEBXI@Z(ptr long) _ValidateRead
+@ cdecl -arch=win32 ?_ValidateWrite@@YAHPAXI@Z(ptr long) _ValidateWrite
+@ cdecl -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z(ptr long) _ValidateWrite
 @ cdecl __uncaught_exception()
-@ stub ?_inconsistency@@YAXXZ
+@ cdecl ?_inconsistency@@YAXXZ() _inconsistency
 @ cdecl -arch=win32 ?_invalid_parameter@@YAXPBG00II@Z(wstr wstr wstr long long) _invalid_parameter
 @ cdecl -arch=win64 ?_invalid_parameter@@YAXPEBG00I_K@Z(wstr wstr wstr long long) _invalid_parameter
 @ cdecl -arch=win32 ?_is_exception_typeof@@YAHABVtype_info@@PAU_EXCEPTION_POINTERS@@@Z(ptr ptr) _is_exception_typeof
@@ -113,20 +113,20 @@
 @ cdecl -arch=win32 ?_query_new_handler@@YAP6AHI@ZXZ() _query_new_handler
 @ cdecl -arch=win64 ?_query_new_handler@@YAP6AH_K@ZXZ() _query_new_handler
 @ cdecl ?_query_new_mode@@YAHXZ() _query_new_mode
-@ stub -arch=win32 ?_set_new_handler@@YAP6AHI@ZH@Z  # int (__cdecl*__cdecl _set_new_handler(int))(unsigned int)
-@ stub -arch=win64 ?_set_new_handler@@YAP6AH_K@ZH@Z  # int (__cdecl*__cdecl _set_new_handler(int))(unsigned __int64)
+@ cdecl -arch=win32 ?_set_new_handler@@YAP6AHI@ZH@Z(long) _set_new_handler_zero
+@ cdecl -arch=win64 ?_set_new_handler@@YAP6AH_K@ZH@Z(long) _set_new_handler_zero
 @ cdecl -arch=win32 ?_set_new_handler@@YAP6AHI@ZP6AHI@Z@Z(ptr) _set_new_handler
 @ cdecl -arch=win64 ?_set_new_handler@@YAP6AH_K@ZP6AH0@Z@Z(ptr) _set_new_handler
 @ cdecl ?_set_new_mode@@YAHH@Z(long) _set_new_mode
-@ stub -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZH@Z  # void (__cdecl*__cdecl _set_se_translator(int))(unsigned int,struct _EXCEPTION_POINTERS *)
-@ stub -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZH@Z  # void (__cdecl*__cdecl _set_se_translator(int))(unsigned int,struct _EXCEPTION_POINTERS * __ptr64)
+@ cdecl -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZH@Z(long) _set_se_translator_zero
+@ cdecl -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZH@Z(long) _set_se_translator_zero
 @ cdecl -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z(ptr) _set_se_translator
 @ cdecl -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z(ptr) _set_se_translator
 @ cdecl -arch=win32 ?_sopen@@YAHPBDHHH@Z(str long long long) _sopen
 @ cdecl -arch=win64 ?_sopen@@YAHPEBDHHH@Z(str long long long) _sopen
-@ stub -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ
-@ stub -arch=i386 ?_type_info_dtor_internal_method@type_info@@QAEXXZ  # public: void __thiscall type_info::_type_info_dtor_internal_method(void)
-@ stub -arch=win64 ?_type_info_dtor_internal_method@type_info@@QEAAXXZ  # public: void __cdecl type_info::_type_info_dtor_internal_method(void) __ptr64
+@ cdecl -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ(ptr) type_info_dtor
+@ thiscall -arch=i386 ?_type_info_dtor_internal_method@type_info@@QAEXXZ(ptr) type_info_dtor
+@ cdecl -arch=win64 ?_type_info_dtor_internal_method@type_info@@QEAAXXZ(ptr) type_info_dtor
 @ cdecl -arch=win32 ?_wopen@@YAHPB_WHH@Z(wstr long long) _wopen
 @ cdecl -arch=win64 ?_wopen@@YAHPEB_WHH@Z(wstr long long) _wopen
 @ cdecl -arch=win32 ?_wsopen@@YAHPB_WHHH@Z(wstr long long long) _wsopen
@@ -244,16 +244,16 @@
 @ cdecl __dllonexit(ptr ptr ptr)
 @ cdecl __doserrno()
 @ cdecl __dstbias() __p__dstbias
-@ stub -arch=i386 ___fls_getvalue@4
-@ stub -arch=x86_64 __fls_getvalue
-@ stub -arch=i386 ___fls_setvalue@8
-@ stub -arch=x86_64 __fls_setvalue
+@ stdcall -arch=i386 ___fls_getvalue@4(long) fls_getvalue_stdcall
+@ cdecl -arch=x86_64 __fls_getvalue(long)
+@ stdcall -arch=i386 ___fls_setvalue@8(long ptr) fls_setvalue_stdcall
+@ cdecl -arch=x86_64 __fls_setvalue(long ptr)
 @ cdecl __fpecode()
 @ cdecl __free_locale(ptr) _free_locale
-@ stub __get_app_type
+@ cdecl __get_app_type()
 @ cdecl __get_current_locale() _get_current_locale
-@ stub __get_flsindex
-@ stub __get_tlsindex
+@ cdecl __get_flsindex() __get_tlsindex
+@ cdecl __get_tlsindex()
 @ cdecl __getmainargs(ptr ptr ptr long ptr)
 @ extern __initenv MSVCRT___initenv
 @ cdecl __iob_func()
@@ -306,7 +306,7 @@
 @ cdecl -arch=i386 __p__iob() __iob_func
 @ stub __p__mbcasemap()
 @ cdecl __p__mbctype()
-@ stub __p__osplatform
+@ cdecl __p__osplatform()
 @ cdecl __p__osver()
 @ cdecl __p__pctype()
 @ cdecl __p__pgmptr()
@@ -325,7 +325,7 @@
 @ cdecl __pxcptinfoptrs()
 @ cdecl __report_gsfailure(ptr)
 @ cdecl __set_app_type(long)
-@ stub __set_flsgetvalue
+@ cdecl __set_flsgetvalue()
 @ extern __setlc_active MSVCRT___setlc_active
 @ cdecl __setusermatherr(ptr)
 @ cdecl __strncnt(str long)
@@ -457,7 +457,7 @@
 @ cdecl _decode_pointer(ptr)
 @ cdecl _difftime32(long long)
 @ cdecl _difftime64(int64 int64)
-@ stub _dosmaperr
+@ cdecl _dosmaperr(long)
 @ extern _dstbias MSVCRT__dstbias
 @ cdecl _dup(long)
 @ cdecl _dup2(long long)
@@ -583,7 +583,7 @@
 @ cdecl _get_unexpected()
 @ cdecl _get_winmajor(ptr)
 @ cdecl _get_winminor(ptr)
-@ stub _get_winver
+@ cdecl _get_winver(ptr)
 @ cdecl _get_wpgmptr(ptr)
 @ cdecl _getc_nolock(ptr) _fgetc_nolock
 @ cdecl _getch()
@@ -592,7 +592,7 @@
 @ cdecl _getche_nolock()
 @ cdecl _getcwd(str long)
 @ cdecl _getdcwd(long str long)
-@ stub _getdcwd_nolock
+@ cdecl _getdcwd_nolock(long ptr long)
 @ cdecl _getdiskfree(long ptr)
 @ cdecl _getdllprocaddr(long str long)
 @ cdecl _getdrive()
@@ -1262,7 +1262,7 @@
 @ cdecl _wfullpath(ptr wstr long)
 @ cdecl _wgetcwd(wstr long)
 @ cdecl _wgetdcwd(long wstr long)
-@ stub _wgetdcwd_nolock
+@ cdecl _wgetdcwd_nolock(long ptr long)
 @ cdecl _wgetenv(wstr)
 @ cdecl _wgetenv_s(ptr ptr long wstr)
 @ extern _winmajor MSVCRT__winmajor
diff --git a/dlls/msvcr90/msvcr90.spec b/dlls/msvcr90/msvcr90.spec
index b83854b..ee607ae 100644
--- a/dlls/msvcr90/msvcr90.spec
+++ b/dlls/msvcr90/msvcr90.spec
@@ -85,22 +85,22 @@
 @ cdecl -arch=win64 ??_U@YAPEAX_KHPEBDH@Z(long long str long) operator_new_dbg
 @ cdecl -arch=win32 ??_V@YAXPAX@Z(ptr) operator_delete
 @ cdecl -arch=win64 ??_V@YAXPEAX@Z(ptr) operator_delete
-@ stub -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z  # private: static char const * __cdecl type_info::_Name_base(class type_info const *,struct __type_info_node *)
-@ stub -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z  # private: static char const * __ptr64 __cdecl type_info::_Name_base(class type_info const * __ptr64,struct __type_info_node * __ptr64)
-@ stub -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z  # private: static char const * __cdecl type_info::_Name_base_internal(class type_info const *,struct __type_info_node *)
-@ stub -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z  # private: static char const * __ptr64 __cdecl type_info::_Name_base_internal(class type_info const * __ptr64,struct __type_info_node * __ptr64)
-@ stub -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z  # private: static void __cdecl type_info::_Type_info_dtor(class type_info *)
-@ stub -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z  # private: static void __cdecl type_info::_Type_info_dtor(class type_info * __ptr64)
-@ stub -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z  # private: static void __cdecl type_info::_Type_info_dtor_internal(class type_info *)
-@ stub -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z  # private: static void __cdecl type_info::_Type_info_dtor_internal(class type_info * __ptr64)
-@ stub -arch=win32 ?_ValidateExecute@@YAHP6GHXZ@Z  # int __cdecl _ValidateExecute(int (__stdcall*)(void))
-@ stub -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z  # int __cdecl _ValidateExecute(__int64 (__cdecl*)(void))
-@ stub -arch=win32 ?_ValidateRead@@YAHPBXI@Z  # int __cdecl _ValidateRead(void const *,unsigned int)
-@ stub -arch=win64 ?_ValidateRead@@YAHPEBXI@Z  # int __cdecl _ValidateRead(void const * __ptr64,unsigned int)
-@ stub -arch=win32 ?_ValidateWrite@@YAHPAXI@Z  # int __cdecl _ValidateWrite(void *,unsigned int)
-@ stub -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z  # int __cdecl _ValidateWrite(void * __ptr64,unsigned int)
+@ cdecl -arch=win32 ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win64 ?_Name_base@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win32 ?_Name_base_internal@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win64 ?_Name_base_internal@type_info@@CAPEBDPEBV1@PEAU__type_info_node@@@Z(ptr ptr) type_info__Name_base
+@ cdecl -arch=win32 ?_Type_info_dtor@type_info@@CAXPAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win64 ?_Type_info_dtor@type_info@@CAXPEAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win32 ?_Type_info_dtor_internal@type_info@@CAXPAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win64 ?_Type_info_dtor_internal@type_info@@CAXPEAV1@@Z(ptr) type_info__Type_info_dtor
+@ cdecl -arch=win32 ?_ValidateExecute@@YAHP6GHXZ@Z(ptr) _ValidateExecute
+@ cdecl -arch=win64 ?_ValidateExecute@@YAHP6A_JXZ@Z(ptr) _ValidateExecute
+@ cdecl -arch=win32 ?_ValidateRead@@YAHPBXI@Z(ptr long) _ValidateRead
+@ cdecl -arch=win64 ?_ValidateRead@@YAHPEBXI@Z(ptr long) _ValidateRead
+@ cdecl -arch=win32 ?_ValidateWrite@@YAHPAXI@Z(ptr long) _ValidateWrite
+@ cdecl -arch=win64 ?_ValidateWrite@@YAHPEAXI@Z(ptr long) _ValidateWrite
 @ cdecl __uncaught_exception()
-@ stub ?_inconsistency@@YAXXZ
+@ cdecl ?_inconsistency@@YAXXZ() _inconsistency
 @ cdecl -arch=win32 ?_invalid_parameter@@YAXPBG00II@Z(wstr wstr wstr long long) _invalid_parameter
 @ cdecl -arch=win64 ?_invalid_parameter@@YAXPEBG00I_K@Z(wstr wstr wstr long long) _invalid_parameter
 @ cdecl -arch=win32 ?_is_exception_typeof@@YAHABVtype_info@@PAU_EXCEPTION_POINTERS@@@Z(ptr ptr) _is_exception_typeof
@@ -113,20 +113,20 @@
 @ cdecl -arch=win32 ?_query_new_handler@@YAP6AHI@ZXZ() _query_new_handler
 @ cdecl -arch=win64 ?_query_new_handler@@YAP6AH_K@ZXZ() _query_new_handler
 @ cdecl ?_query_new_mode@@YAHXZ() _query_new_mode
-@ stub -arch=win32 ?_set_new_handler@@YAP6AHI@ZH@Z  # int (__cdecl*__cdecl _set_new_handler(int))(unsigned int)
-@ stub -arch=win64 ?_set_new_handler@@YAP6AH_K@ZH@Z  # int (__cdecl*__cdecl _set_new_handler(int))(unsigned __int64)
+@ cdecl -arch=win32 ?_set_new_handler@@YAP6AHI@ZH@Z(long) _set_new_handler_zero
+@ cdecl -arch=win64 ?_set_new_handler@@YAP6AH_K@ZH@Z(long) _set_new_handler_zero
 @ cdecl -arch=win32 ?_set_new_handler@@YAP6AHI@ZP6AHI@Z@Z(ptr) _set_new_handler
 @ cdecl -arch=win64 ?_set_new_handler@@YAP6AH_K@ZP6AH0@Z@Z(ptr) _set_new_handler
 @ cdecl ?_set_new_mode@@YAHH@Z(long) _set_new_mode
-@ stub -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZH@Z  # void (__cdecl*__cdecl _set_se_translator(int))(unsigned int,struct _EXCEPTION_POINTERS *)
-@ stub -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZH@Z  # void (__cdecl*__cdecl _set_se_translator(int))(unsigned int,struct _EXCEPTION_POINTERS * __ptr64)
+@ cdecl -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZH@Z(long) _set_se_translator_zero
+@ cdecl -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZH@Z(long) _set_se_translator_zero
 @ cdecl -arch=win32 ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z(ptr) _set_se_translator
 @ cdecl -arch=win64 ?_set_se_translator@@YAP6AXIPEAU_EXCEPTION_POINTERS@@@ZP6AXI0@Z@Z(ptr) _set_se_translator
 @ cdecl -arch=win32 ?_sopen@@YAHPBDHHH@Z(str long long long) _sopen
 @ cdecl -arch=win64 ?_sopen@@YAHPEBDHHH@Z(str long long long) _sopen
-@ stub -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ
-@ stub -arch=i386 ?_type_info_dtor_internal_method@type_info@@QAEXXZ  # public: void __thiscall type_info::_type_info_dtor_internal_method(void)
-@ stub -arch=win64 ?_type_info_dtor_internal_method@type_info@@QEAAXXZ  # public: void __cdecl type_info::_type_info_dtor_internal_method(void) __ptr64
+@ cdecl -arch=arm ?_type_info_dtor_internal_method@type_info@@QAAXXZ(ptr) type_info_dtor
+@ thiscall -arch=i386 ?_type_info_dtor_internal_method@type_info@@QAEXXZ(ptr) type_info_dtor
+@ cdecl -arch=win64 ?_type_info_dtor_internal_method@type_info@@QEAAXXZ(ptr) type_info_dtor
 @ cdecl -arch=win32 ?_wopen@@YAHPB_WHH@Z(wstr long long) _wopen
 @ cdecl -arch=win64 ?_wopen@@YAHPEB_WHH@Z(wstr long long) _wopen
 @ cdecl -arch=win32 ?_wsopen@@YAHPB_WHHH@Z(wstr long long long) _wsopen
@@ -235,14 +235,14 @@
 @ cdecl __dllonexit(ptr ptr ptr)
 @ cdecl __doserrno()
 @ cdecl __dstbias() __p__dstbias
-@ stub ___fls_getvalue@4
-@ stub ___fls_setvalue@8
+@ stdcall ___fls_getvalue@4(long) fls_getvalue_stdcall
+@ stdcall ___fls_setvalue@8(long ptr) fls_setvalue_stdcall
 @ cdecl __fpecode()
 @ cdecl __free_locale(ptr) _free_locale
-@ stub __get_app_type
+@ cdecl __get_app_type()
 @ cdecl __get_current_locale() _get_current_locale
-@ stub __get_flsindex
-@ stub __get_tlsindex
+@ cdecl __get_flsindex() __get_tlsindex
+@ cdecl __get_tlsindex()
 @ cdecl __getmainargs(ptr ptr ptr long ptr)
 @ extern __initenv MSVCRT___initenv
 @ cdecl __iob_func()
@@ -309,7 +309,7 @@
 @ cdecl __pxcptinfoptrs()
 @ cdecl __report_gsfailure(ptr)
 @ cdecl __set_app_type(long)
-@ stub __set_flsgetvalue
+@ cdecl __set_flsgetvalue()
 @ extern __setlc_active MSVCRT___setlc_active
 @ cdecl __setusermatherr(ptr)
 @ cdecl __strncnt(str long)
@@ -440,7 +440,7 @@
 @ cdecl _decode_pointer(ptr)
 @ cdecl _difftime32(long long)
 @ cdecl _difftime64(int64 int64)
-@ stub _dosmaperr
+@ cdecl _dosmaperr(long)
 @ extern _dstbias MSVCRT__dstbias
 @ cdecl _dup(long)
 @ cdecl _dup2(long long)
@@ -570,7 +570,7 @@
 @ cdecl _getche_nolock()
 @ cdecl _getcwd(str long)
 @ cdecl _getdcwd(long str long)
-@ stub _getdcwd_nolock
+@ cdecl _getdcwd_nolock(long ptr long)
 @ cdecl _getdiskfree(long ptr)
 @ cdecl _getdllprocaddr(long str long)
 @ cdecl _getdrive()
@@ -1237,7 +1237,7 @@
 @ cdecl _wfullpath(ptr wstr long)
 @ cdecl _wgetcwd(wstr long)
 @ cdecl _wgetdcwd(long wstr long)
-@ stub _wgetdcwd_nolock
+@ cdecl _wgetdcwd_nolock(long ptr long)
 @ cdecl _wgetenv(wstr)
 @ cdecl _wgetenv_s(ptr ptr long wstr)
 @ cdecl _wmakepath(ptr wstr wstr wstr wstr)
diff --git a/dlls/msvcrt/cpp.c b/dlls/msvcrt/cpp.c
index 99473fd..faf1ce7 100644
--- a/dlls/msvcrt/cpp.c
+++ b/dlls/msvcrt/cpp.c
@@ -989,6 +989,32 @@ const char * __thiscall type_info_name_internal_method(type_info * _this, struct
     return type_info_name(_this);
 }
 
+/* ?_Name_base@type_info@@CAPBDPBV1@PAU__type_info_node@@@Z and _Name_base_internal */
+const char * __cdecl type_info__Name_base(type_info *ti, struct __type_info_node *node)
+{
+    return type_info_name_internal_method(ti, node);
+}
+
+/* ?_Type_info_dtor@type_info@@CAXPAV1@@Z and _Type_info_dtor_internal */
+void __cdecl type_info__Type_info_dtor(type_info *ti)
+{
+    type_info_dtor(ti);
+}
+
+/* ?_inconsistency@@YAXXZ: the exception handling data is inconsistent */
+void __cdecl _inconsistency(void)
+{
+    ERR("inconsistent exception handling data\n");
+    terminate();
+}
+
+/* ?_set_se_translator@@YAP6AXIPAU_EXCEPTION_POINTERS@@@ZH@Z: the overload of _set_se_translator(0) */
+_se_translator_function __cdecl _set_se_translator_zero(int zero)
+{
+    if (zero) FIXME("(%d) not 0\n", zero);
+    return _set_se_translator(NULL);
+}
+
 void* __cdecl __AdjustPointer(void *obj, const this_ptr_offsets *off)
 {
     return get_this_pointer(off, obj);
diff --git a/dlls/msvcrt/data.c b/dlls/msvcrt/data.c
index 6e764e5..84dd13e 100644
--- a/dlls/msvcrt/data.c
+++ b/dlls/msvcrt/data.c
@@ -590,6 +590,44 @@ void CDECL __set_app_type(int app_type)
   MSVCRT_app_type = app_type;
 }
 
+/*********************************************************************
+ *		__get_app_type (MSVCRT.@)
+ */
+int CDECL __get_app_type(void)
+{
+  return MSVCRT_app_type;
+}
+
+#if _MSVCR_VER>=140
+/*********************************************************************
+ *		_query_app_type (UCRTBASE.@)
+ */
+int CDECL _query_app_type(void)
+{
+  return MSVCRT_app_type;
+}
+#endif
+
+#if _MSVCR_VER>=80
+/*********************************************************************
+ *		__p__osplatform (MSVCR80.@)
+ */
+unsigned int * CDECL __p__osplatform(void)
+{
+  return &MSVCRT__osplatform;
+}
+
+/*********************************************************************
+ *		_get_winver (MSVCR80.@)
+ */
+int CDECL _get_winver(unsigned int *value)
+{
+  if (!MSVCRT_CHECK_PMT(value != NULL)) return EINVAL;
+  *value = MSVCRT__winver;
+  return 0;
+}
+#endif
+
 #if _MSVCR_VER>=140
 
 /*********************************************************************
diff --git a/dlls/msvcrt/dir.c b/dlls/msvcrt/dir.c
index 1d91d5f..1cf2b0d 100644
--- a/dlls/msvcrt/dir.c
+++ b/dlls/msvcrt/dir.c
@@ -1898,3 +1898,21 @@ void CDECL _wsearchenv(const wchar_t* file, const wchar_t* env, wchar_t *buf)
 {
     _wsearchenv_s(file, env, buf, MAX_PATH);
 }
+
+#if _MSVCR_VER >= 80
+/*********************************************************************
+ *		_getdcwd_nolock (MSVCR80.@)
+ */
+char * CDECL _getdcwd_nolock(int drive, char *buf, int size)
+{
+    return _getdcwd(drive, buf, size);
+}
+
+/*********************************************************************
+ *		_wgetdcwd_nolock (MSVCR80.@)
+ */
+wchar_t * CDECL _wgetdcwd_nolock(int drive, wchar_t *buf, int size)
+{
+    return _wgetdcwd(drive, buf, size);
+}
+#endif
diff --git a/dlls/msvcrt/exception_ptr.c b/dlls/msvcrt/exception_ptr.c
index 708a612..aab4e0d 100644
--- a/dlls/msvcrt/exception_ptr.c
+++ b/dlls/msvcrt/exception_ptr.c
@@ -110,6 +110,24 @@ void __cdecl __ExceptionPtrCopy(exception_ptr *ep, const exception_ptr *copy)
         InterlockedIncrement(copy->ref);
 }
 
+#if _MSVCR_VER >= 110
+/*********************************************************************
+ * ?__ExceptionPtrSwap@@YAXPAX0@Z
+ * ?__ExceptionPtrSwap@@YAXPEAX0@Z
+ */
+void __cdecl __ExceptionPtrSwap(exception_ptr *a, exception_ptr *b)
+{
+    exception_ptr t;
+
+    TRACE("(%p %p)\n", a, b);
+#ifndef SG_MUTANT_EXCEPTIONPTR_SWAP_NOP
+    t = *a;
+    *a = *b;
+    *b = t;
+#endif
+}
+#endif
+
 /*********************************************************************
  * ?__ExceptionPtrAssign@@YAXPAXPBX@Z
  * ?__ExceptionPtrAssign@@YAXPEAXPEBX@Z
diff --git a/dlls/msvcrt/heap.c b/dlls/msvcrt/heap.c
index e7eb8e1..7251060 100644
--- a/dlls/msvcrt/heap.c
+++ b/dlls/msvcrt/heap.c
@@ -957,3 +957,27 @@ void msvcrt_destroy_heap(void)
     if(sb_heap)
         HeapDestroy(sb_heap);
 }
+
+#if _MSVCR_VER >= 80
+/*********************************************************************
+ * ?_set_new_handler@@YAP6AHI@ZH@Z (MSVCR80.@)
+ *
+ * The overload of _set_new_handler(0).
+ */
+MSVCRT_new_handler_func CDECL _set_new_handler_zero(int zero)
+{
+    if (zero) FIXME("(%d) not 0\n", zero);
+    return _set_new_handler(NULL);
+}
+#endif
+
+#if _MSVCR_VER >= 110 && _MSVCR_VER <= 120
+/*********************************************************************
+ * ??3@YAXPAXHPBDH@Z (MSVCR110.@) operator delete(void*, int, const char*, int)
+ * ??_V@YAXPAXHPBDH@Z (MSVCR110.@) operator delete[] of the same form
+ */
+void CDECL operator_delete_dbg(void *mem, int type, const char *file, int line)
+{
+    free(mem);
+}
+#endif
diff --git a/dlls/msvcrt/misc.c b/dlls/msvcrt/misc.c
index fcfbcda..40a6502 100644
--- a/dlls/msvcrt/misc.c
+++ b/dlls/msvcrt/misc.c
@@ -801,3 +801,85 @@ void* CDECL _lsearch_s(const void *match, void *start, unsigned int *array_size,
     return found;
 }
 #endif
+
+#if _MSVCR_VER >= 80
+/*********************************************************************
+ * _dosmaperr (MSVCR80.@)
+ *
+ * errno and _doserrno from a system error
+ */
+void CDECL _dosmaperr(DWORD err)
+{
+    msvcrt_set_errno(err);
+}
+
+/*********************************************************************
+ * __get_tlsindex (MSVCR80.@)
+ * __get_flsindex (MSVCR80.@)
+ */
+DWORD CDECL __get_tlsindex(void)
+{
+    return msvcrt_tls_index;
+}
+
+/*********************************************************************
+ * __fls_getvalue (MSVCR80.@)
+ */
+void * CDECL __fls_getvalue(DWORD index)
+{
+    return FlsGetValue(index);
+}
+
+/*********************************************************************
+ * __fls_setvalue (MSVCR80.@)
+ */
+BOOL CDECL __fls_setvalue(DWORD index, void *value)
+{
+    return FlsSetValue(index, value);
+}
+
+/* ___fls_getvalue@4, ___fls_setvalue@8 */
+void * WINAPI fls_getvalue_stdcall(DWORD index)
+{
+    return FlsGetValue(index);
+}
+
+BOOL WINAPI fls_setvalue_stdcall(DWORD index, void *value)
+{
+    return FlsSetValue(index, value);
+}
+
+/*********************************************************************
+ * __set_flsgetvalue (MSVCR80.@)
+ *
+ * The function that reads a fiber local value.
+ */
+void * CDECL __set_flsgetvalue(void)
+{
+    return fls_getvalue_stdcall;
+}
+
+/*********************************************************************
+ * ?_ValidateRead@@YAHPBXI@Z (MSVCR80.@)
+ */
+int CDECL _ValidateRead(const void *p, unsigned int size)
+{
+    return p && !IsBadReadPtr(p, size);
+}
+
+/*********************************************************************
+ * ?_ValidateWrite@@YAHPAXI@Z (MSVCR80.@)
+ */
+int CDECL _ValidateWrite(void *p, unsigned int size)
+{
+    return p && !IsBadWritePtr(p, size);
+}
+
+/*********************************************************************
+ * ?_ValidateExecute@@YAHP6GHXZ@Z (MSVCR80.@)
+ */
+int CDECL _ValidateExecute(FARPROC f)
+{
+    return f && !IsBadCodePtr(f);
+}
+#endif
diff --git a/dlls/msvcrt/msvcrt.spec b/dlls/msvcrt/msvcrt.spec
index ca1f35a..4ff185e 100644
--- a/dlls/msvcrt/msvcrt.spec
+++ b/dlls/msvcrt/msvcrt.spec
@@ -213,7 +213,7 @@
 @ cdecl __doserrno()
 @ cdecl __dstbias() __p__dstbias
 @ cdecl __fpecode()
-@ stub __get_app_type
+@ cdecl __get_app_type()
 @ cdecl __getmainargs(ptr ptr ptr long ptr)
 @ extern __initenv MSVCRT___initenv
 @ cdecl __iob_func()
diff --git a/dlls/ucrtbase/ucrtbase.spec b/dlls/ucrtbase/ucrtbase.spec
index 87bb833..9469254 100644
--- a/dlls/ucrtbase/ucrtbase.spec
+++ b/dlls/ucrtbase/ucrtbase.spec
@@ -1858,7 +1858,7 @@
 @ cdecl _putwch(long)
 @ cdecl _putwch_nolock(long)
 @ cdecl _putws(wstr)
-@ stub _query_app_type
+@ cdecl _query_app_type()
 @ cdecl _query_new_handler()
 @ cdecl _query_new_mode()
 @ cdecl _read(long ptr long)
